common/src/main/java/io/pravega/common/util/BitConverter.java

```diff
@@ -226,4 +226,57 @@ public static long readLong(byte[] source, int position) {
                 | (source[position + 6] & 0xFF) << 8
                 | (source[position + 7] & 0xFF);
     }
+
+    /**
+     * Writes the given 64-bit Unsigned Long to the given byte array at the given offset. This value can then be
+     * deserialized using {@link #readUnsignedLong}. This method is not interoperable with {@link #readLong}.
+     *
+     * The advantage of serializing as Unsigned Long (vs. a normal Signed Long) is that the serialization will have the
+     * same natural order as the input value type (i.e., if compared using a lexicographic bitwise comparator such as
+     * ByteArrayComparator, it will have the same ordering as the typical Long type).
+     *
+     * @param target The byte array to write to.
+     * @param offset The offset within the byte array to write at.
+     * @param value  The (signed) value to write. The value will be converted into the range [0, 2^64-1] before
+     *               serialization: if negative, then {@link Long#MAX_VALUE} + 1 will be added to it, otherwise the
+     *               first bit will be set to 1.
+     * @return The number of bytes written.
+     */
+    public static int writeUnsignedLong(byte[] target, int offset, long value) {
```

Comment:
It does more than that. When serialized like this, doing a bitwise comparison of the resulting output will order the values in the same order as the inputs - mimics a natural ordering. Simply flipping the high order bits will not achieve this because negative values are encoded using 2's complement.

Is there an improvement you are suggesting?

Reply:
I'm saying you are already doing this. 
value + Long.MAX_VALUE + 1 is equal to value + Long.MIN_VALUE which if the high order bit in known to be one, which your if condition covers, means the addition allways rolls over, which means you have just flipped the highest bit. The else block also just flips the highest bit because the highest bit is guarenteed to be a 0.

Reply:
I got it now. I misunderstood what you meant originally. I simplified it.